controllers/vspherecluster_controller.go

```diff
@@ -290,56 +455,118 @@ func (r clusterReconciler) reconcileControlPlaneEndpoint(ctx *context.ClusterCon
+			return false, errors.Wrapf(err,
+				"failed to get preferred IP address for VSphereMachine %s %s/%s",
+				vsphereMachine.GroupVersionKind(),
+				vsphereMachine.Namespace,
+				vsphereMachine.Name)
 		}
 
 		// Set the ControlPlaneEndpoint so the CAPI controller can read the
 		// value into the analogous CAPI Cluster using an UnstructuredReader.
 		ctx.VSphereCluster.Spec.ControlPlaneEndpoint.Host = ipAddr
-		ctx.VSphereCluster.Spec.ControlPlaneEndpoint.Port = apiEndpointPort
-
-		// Enqueue a reconcile request for the cluster when the target API
-		// server is online.
-		go func() {
-			// Block until the target API server is online.
-			wait.PollImmediateInfinite(time.Second*1, func() (bool, error) { return r.isAPIServerOnline(ctx), nil })
-			ctx.Logger.Info("triggering GenericEvent", "reason", "api-server-online")
-			eventChannel := ctx.GetGenericEventChannelFor(ctx.VSphereCluster.GetObjectKind().GroupVersionKind())
-			eventChannel <- event.GenericEvent{
-				Meta:   ctx.VSphereCluster,
-				Object: ctx.VSphereCluster,
-			}
-		}()
+		ctx.VSphereCluster.Spec.ControlPlaneEndpoint.Port = defaultAPIEndpointPort
+		ctx.Logger.Info(
+			"ControlPlaneEndpoin discovered via control plane machine",
+			"controlPlaneEndpoint", ctx.VSphereCluster.Spec.ControlPlaneEndpoint)
+		return true, nil
+	}
 
-		ctx.Logger.V(4).Info(
-			"found API endpoint via control plane machine",
-			"host", ctx.VSphereCluster.Spec.ControlPlaneEndpoint.Host,
-			"port", ctx.VSphereCluster.Spec.ControlPlaneEndpoint.Port)
+	return false, errors.Errorf("unable to determine control plane endpoint for %s", ctx)
+}
 
-		return nil
-	}
-	return infrautilv1.ErrNoMachineIPAddr
+var (
+	// apiServerTriggers is used to prevent multiple goroutines for a single
+	// Cluster that poll to see if the target API server is online.
+	apiServerTriggers   = map[types.UID]struct{}{}
+	apiServerTriggersMu sync.Mutex
+)
+
+func (r clusterReconciler) reconcileVSphereClusterWhenAPIServerIsOnline(ctx *context.ClusterContext) {
+	if ctx.Cluster.Status.ControlPlaneInitialized {
+		ctx.Logger.Info("skipping reconcile when API server is online",
+			"reason", "controlPlaneInitialized")
+		return
+	}
+	apiServerTriggersMu.Lock()
+	defer apiServerTriggersMu.Unlock()
+	if _, ok := apiServerTriggers[ctx.Cluster.UID]; ok {
+		ctx.Logger.Info("skipping reconcile when API server is online",
+			"reason", "alreadyPolling")
+		return
+	}
+	apiServerTriggers[ctx.Cluster.UID] = struct{}{}
+	go func() {
+		// Block until the target API server is online.
+		ctx.Logger.Info("start polling API server for online check")
+		wait.PollImmediateInfinite(time.Second*1, func() (bool, error) { return r.isAPIServerOnline(ctx), nil })
+		ctx.Logger.Info("stop polling API server for online check")
+		ctx.Logger.Info("triggering GenericEvent", "reason", "api-server-online")
+		eventChannel := ctx.GetGenericEventChannelFor(ctx.VSphereCluster.GetObjectKind().GroupVersionKind())
```

Comment:
Hi @yastij,

Because it's still required for deployments without an LB. This:

* Causes the `VSphereCluster` to set its `ControlPlaneEndpoint` from a `VSphereMachine`
* Which then causes the CAPI `Cluster` to reconcile since it watches its infrastructure cluster.